src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.Executor.cs

```diff
@@ -162,6 +162,89 @@ public IEnumerable<DiagnosticData> ConvertToLocalDiagnostics(Document targetDocu
                 return await MergeProjectDiagnosticAnalyzerDiagnosticsAsync(project, stateSets, analyzerDriverOpt?.Compilation, result, cancellationToken).ConfigureAwait(false);
             }
 
+            private async Task<ImmutableDictionary<DiagnosticAnalyzer, AnalysisResult>> ComputeDiagnosticsAsync(
+                CompilationWithAnalyzers analyzerDriverOpt, Project project, IEnumerable<StateSet> stateSets,
+                ImmutableDictionary<DiagnosticAnalyzer, AnalysisResult> existing, CancellationToken cancellationToken)
+            {
+                // PERF: check whether we can reduce number of analyzers we need to run.
+                //       this can happen since caller could have created the driver with different set of analyzers that are different
+                //       than what we used to create the cache.
+                var version = await GetDiagnosticVersionAsync(project, cancellationToken).ConfigureAwait(false);
+                var analyzersToRun = ReduceAnalyzersToRun(analyzerDriverOpt, version, existing);
+                if (analyzersToRun != null)
+                {
+                    // it looks like we can reduce the set. create new CompilationWithAnalyzer.
+                    var analyzerDriverWithReducedSet = await _owner._compilationManager.CreateAnalyzerDriverAsync(
+                        project, analyzersToRun.Value, analyzerDriverOpt.AnalysisOptions.ReportSuppressedDiagnostics, cancellationToken).ConfigureAwait(false);
+
+                    var result = await ComputeDiagnosticsAsync(analyzerDriverWithReducedSet, project, stateSets, cancellationToken).ConfigureAwait(false);
+                    return MergeExistingDiagnostics(version, existing, result);
+                }
+
+                // we couldn't reduce the set.
+                return await ComputeDiagnosticsAsync(analyzerDriverOpt, project, stateSets, cancellationToken).ConfigureAwait(false);
+            }
+
+            private ImmutableDictionary<DiagnosticAnalyzer, AnalysisResult> MergeExistingDiagnostics(
+                VersionStamp version, ImmutableDictionary<DiagnosticAnalyzer, AnalysisResult> existing, ImmutableDictionary<DiagnosticAnalyzer, AnalysisResult> result)
+            {
+                // quick bail out.
+                if (existing.IsEmpty)
+                {
+                    return result;
+                }
+
+                foreach (var kv in existing)
+                {
+                    if (kv.Value.Version != version)
+                    {
+                        continue;
+                    }
+
+                    result = result.SetItem(kv.Key, kv.Value);
+                }
+
+                return result;
+            }
+
+            private ImmutableArray<DiagnosticAnalyzer>? ReduceAnalyzersToRun(
```

Comment:
Change it to return bool with ImmutableArray<DiagnosticAnalyzer> as an out param